Source/WebCore/html/parser/HTMLDocumentParserFastPath.cpp

```diff
@@ -0,0 +1,1053 @@
+#include "HTMLBRElement.h"
+#include "HTMLButtonElement.h"
+#include "HTMLDivElement.h"
+#include "HTMLEntityParser.h"
+#include "HTMLInputElement.h"
+#include "HTMLLIElement.h"
+#include "HTMLLabelElement.h"
+#include "HTMLNameCache.h"
+#include "HTMLNames.h"
+#include "HTMLOListElement.h"
+#include "HTMLParagraphElement.h"
+#include "HTMLSelectElement.h"
+#include "HTMLSpanElement.h"
+#include "HTMLUListElement.h"
+#include "QualifiedName.h"
+#include "Settings.h"
+#include <wtf/Span.h>
+#include <wtf/Vector.h>
+#include <wtf/text/AtomString.h>
+
+namespace WebCore {
+
+// Captures the potential outcomes for fast path html parser.
+// These values are persisted to logs. Entries should not be renumbered and
+// numeric values should never be reused.
+// This is exposed for testing.
+enum class HTMLFastPathResult {
+    Succeeded,
+    FailedTracingEnabled,
+    FailedParserContentPolicy,
+    FailedInForm,
+    FailedUnsupportedContextTag,
+    FailedOptionWithChild,
+    FailedDidntReachEndOfInput,
+    FailedContainsNull,
+    FailedParsingTagName,
+    FailedParsingQuotedAttributeValue,
+    FailedParsingUnquotedAttributeValue,
+    FailedParsingQuotedEscapedAttributeValue,
+    FailedParsingUnquotedEscapedAttributeValue,
+    FailedParsingCharacterReference,
+    FailedEndOfInputReached,
+    FailedParsingAttributes,
+    FailedParsingSpecificElements,
+    FailedParsingElement,
+    FailedUnsupportedTag,
+    FailedEndOfInputReachedForContainer,
+    FailedUnexpectedTagNameCloseState,
+    FailedEndTagNameMismatch,
+    FailedShadowRoots,
+    FailedOnAttribute,
+    FailedMaxDepth,
+    FailedBigText,
+    FailedCssPseudoDirEnabledAndDirAttributeDirty,
+    MaxValue = FailedCssPseudoDirEnabledAndDirAttributeDirty,
+};
+
+template<class Char, size_t n> static bool operator==(Span<const Char> span, const char (&s)[n])
+{
+    if (span.size() != n - 1)
+        return false;
+
+    for (size_t i = 0; i < n - 1; ++i) {
+        if (span[i] != s[i])
+            return false;
+    }
+    return true;
+}
+
+#if ASSERT_ENABLED
+template<size_t n> static constexpr bool onlyContainsLowercaseASCIILetters(const char (&s)[n])
+{
+    for (size_t i = 0; i < n - 1; ++i) {
+        if (s[i] < 'a' || s[i] > 'z')
+            return false;
+    }
+    return true;
+}
+#endif // ASSERT_ENABLED
+
+// A hash function that is just good enough to distinguish the supported tagNames. It needs to be
+// adapted as soon as we have colliding tagNames. The implementation was chosen to map to a dense
+// integer range to allow for compact switch jump-tables. If adding support for a new tag results
+// in a collision, then pick a new function that minimizes the number of operations and results
+// in a dense integer range.
+template<size_t n> static constexpr uint32_t tagNameHash(const char (&s)[n])
+{
+    // The fast-path parser only scans for letters in tagNames.
+    ASSERT(onlyContainsLowercaseASCIILetters<n>(s));
+    ASSERT(s[n - 1] == '\0');
```

Comment:
Surprised that these don’t need to use `ASSERT_IN_CONSTEXPR_CONTEXT`.

Reply:
It does build without it and I admit I hadn't heard of `ASSERT_UNDER_CONSTEXPR_CONTEXT()` before. I'll try switching to `ASSERT_UNDER_CONSTEXPR_CONTEXT()`